git webkit appears to be incorrectly using the first radar link in bugzilla instead of the last one

briannafan opened the pull request:
#### 93524d44866405408cc350115d87c4b5fdfcaa4c
<pre>
git webkit appears to be incorrectly using the first radar link in bugzilla instead of the last one
<a href="https://bugs.webkit.org/show_bug.cgi?id=292230">https://bugs.webkit.org/show_bug.cgi?id=292230</a>
<a href="https://rdar.apple.com/149093470">rdar://149093470</a>

Reviewed by Sam Sneddon and Jonathan Bedard.

We should prioritize bugs commented more recently, as they are likely corrections. Using the first link or
prioritizing the bug importer over a human comment would be incorrect in that case.

To match WKBI, we only match bugs at the start of a line.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate): Reverse the comments and remove preference for the bug importer.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
(TestBugzilla.test_reference_order): Add test for ordering.
(TestBugzilla.test_reference_parse):
(TestBugzilla.test_reference_multiline):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
(TestRadar.test_reference_parse):
(TestRadar.test_reference_multiline):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker): Update REFERENCE_RE to only look for bugs at the start of a line.

Canonical link: <a href="https://commits.webkit.org/294444@main">https://commits.webkit.org/294444@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ed09ea4c4a269061d08a05cc7931f33ba58729d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101859 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/21527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11843 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/107017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/52493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103899 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/21835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/30034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/107017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/52493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104866 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/21835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/91973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/107017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/101332 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/21835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/51845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/21835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/10070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109375 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/28992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/30034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/29353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/88176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23163 "Build is in progress. Recent messages:") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16564 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/28920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/28731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/32054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/30290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -286,20 +286,17 @@ def populate(self, issue, member=None):
-                        continue
-                    issue._references.append(candidate)
-                    refs.add(candidate.link)
+            # In reverse chronological order, prioritize comments with only a bug url, then all comments, then description
```
"only a bug URL" is a notable change from the Radar importer, which always just goes by comments from newest-to-oldest (and the description in Bugzilla is just comment 0!)

gsnedders replied:
I guess there's no reason not to just post the importer logic here:

```python
import re

RADAR_LINK_REGEX = r"^<?(rdar|radr|radar)://(problem/)?(?P<radar_id>\d+)>?.*$"

for comment in reversed(list(comments_for_bug(bugzilla_id))):
    match = re.search(RADAR_LINK_REGEX, comment["text"], re.MULTILINE)
    if match:
        radar_id = match.group("radar_id")
```

Notably this differs insofar as it _requires_ the link to appear at the start of a line, though it allows anything following it.

briannafan replied:
Yeah I think we should follow WKBI. I've updated the regex and logic to reflect this.

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -286,20 +286,17 @@ def populate(self, issue, member=None):
+                refs.add(candidate.link)
 
-            for text in [issue.description] + [comment.content for comment in (issue.comments or []) if comment]:
+            for text in reversed([issue.description] + [comment.content for comment in (issue.comments or []) if comment]):
```
I'd suggest going with chained iterators, and use iterators throughout. From memory, fetching comment contents has a cost for some issue trackers.

Thus, something like:

```python
from itertools import chain

for text in chain(
    (comment.content for comment in (reversed(issue.comments) or ()) if comment),
    (issue.description,),
):
    pass
```

(This also uses tuples and not lists, which is really meaningless optimisation, but it doesn't hurt.)

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
```diff
@@ -29,7 +29,7 @@
 
 class Tracker(object):
-    REFERENCE_RE = re.compile(r'((https|http|rdar|radar)://[^\s><,\'\"}{\]\[)(]*[^\s><,\'\"}{\]\[)(\.\?])')
+    REFERENCE_RE = re.compile(r'^<?((https|http|rdar|radar)://[^\s><,\'\"}{\]\[)(]*[^\s><,\'\"}{\]\[)(\.\?])>?')
```
Pretty sure this needs the `re.MULTILINE` flag to match the WKBI.

We probably want to make sure we have tests both with `<>` and for multiple lines.

briannafan replied:
Good catch! Added flag and tests for multiline - `<>` tests already exist

gsnedders reviewed (approved):
_Ideally_ I'd like to see a test for order *and* multiline, because how those orderings combine is important — but not super vital.

Anyhow, r+! ✨

briannafan committed:
git webkit appears to be incorrectly using the first radar link in bugzilla instead of the last one
https://bugs.webkit.org/show_bug.cgi?id=292230
rdar://149093470

Reviewed by Sam Sneddon and Jonathan Bedard.

We should prioritize bugs commented more recently, as they are likely corrections. Using the first link or
prioritizing the bug importer over a human comment would be incorrect in that case.

To match WKBI, we only match bugs at the start of a line.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate): Reverse the comments and remove preference for the bug importer.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
(TestBugzilla.test_reference_order): Add test for ordering.
(TestBugzilla.test_reference_parse):
(TestBugzilla.test_reference_multiline):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
(TestRadar.test_reference_parse):
(TestRadar.test_reference_multiline):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker): Update REFERENCE_RE to only look for bugs at the start of a line.

Canonical link: https://commits.webkit.org/294444@main
